### ISSUE TITLE
Handling of files that differ only in case on windows (with fsutil file setcasesensitiveinfo enable)

### ISSUE THREAD

**Issue by Rolf-B** [git-for-windows/git#3338]
- [x] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

```
$ git --version --build-options

git version 2.32.0.windows.2
cpu: x86_64
built from commit: 3d45ac813c4adf97fe3733c1f763ab6617d5add5
sizeof-long: 4
sizeof-size_t: 8
shell-path: /bin/sh
feature: fsmonitor--daemon
```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```
$ cmd.exe /c ver

Microsoft Windows [Version 10.0.19041.1110]
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

```
# One of the following:
> type "C:\Program Files\Git\etc\install-options.txt"
> type "C:\Program Files (x86)\Git\etc\install-options.txt"
> type "%USERPROFILE%\AppData\Local\Programs\Git\etc\install-options.txt"
$ cat /etc/install-options.txt

Editor Option: Notepad++
Custom Editor Path: 
Default Branch Option:  
Path Option: Cmd
Plink Path: C:\Program Files\PuTTY\plink.exe
SSH Option: Plink
Tortoise Option: false
CURL Option: OpenSSL
CRLF Option: LFOnly
Bash Terminal Option: MinTTY
Git Pull Behavior Option: Merge
Use Credential Manager: Core
Performance Tweaks FSCache: Enabled
Enable Symlinks: Enabled
Enable Pseudo Console Support: Disabled
Enable FSMonitor: Disabled

```

 - Any other interesting things about your environment that might be related
   to the issue you're seeing?

I have activated the Windows Subsystem for Linux (WSL), which gives me the option to enable case sensitivity for NTFS folders. 
I used  `fsutil file setcasesensitiveinfo D:\projects` enable to create a case sensitive folder for the repository. I didn't do that in previous attempts, and git clone put a [core] ignorecase=true into the repo config. After enabling case sensitivity with FSUTIL, the ignorecase entry didn't appear anymore, so I assume that git close recognized the changed NTFS mode.

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

Git Extensions. I am new to git and totally not familiar with the GIT command line.

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

Since I am still alien to git cmdline, that's difficult.  I could create a minimal testcase like this:
- create a folder and run `fsutil file setcasesensitiveinfo <folder> enable` for it (must activate WSL Feature and be admin for that). This flag is meant for WSL, but it works for Windows Programs, too :). 
- create a personal git repo in this folder
- add 2 textfiles foo.txt and Foo.txt with different content. Check that they both exist and show different content (or you made something wrong with fsutil)
- stage them and commit

 - What did you expect to occur after running these commands?

I did expect that the files are commited and none of them is shown as modified any more.

 - What actually happened instead?

The commit seems to happen, but `git status` shows Foo.txt as modified

 - If the problem was occurring with a specific repository, can you provide the
   URL to that repository to help us with testing?

It's a non-public repo.

**Comment by dscho** [comment]
Since this flag can be enabled on a directory by directory basis, you can end up with a Git worktree that has both a directory respecting case sensitivity as well as another directory that does not.

I am not quite sure that we can ever support such a scenario.

> It's a non-public repo.

This is of course the correct answer. It is equally true that it would be more than just possible to create a new repository and/or script to recreate this problem elsewhere. On the overall time balance sheet, the most reasonable use of time is for the reporter to put in that effort because the alternative would be to put the onus to each and every person who wants to help.

**Comment by Rolf-B** [comment]
You're right, if you don't do it right, you might end up with a cs/ci mix and that would be an impossible mess. That's why I enabled case sensitivity on the parent directory for the repo (D:\projects) and **then** cloned the repo. When you create a subfolder in a case sensitive folder, the new subfolder is case sensitive too. That happened in my case, I checked (fsutil file querycasesensitiveinfo <dir>). And I wouldn't expect from git to support a repo with mixed case sensitivity. Either CI or CS, no mix. Usually, the scenario won't happen. If you create your repo on a CI file system, you just can't produce the problem. And on a CS filesystem like Linux, I don't think that someone intentionally creates foo.txt and Foo.txt. In our repo, I believe that this duplicity is an error and I want to fix it. But I can't, because I don't have Linux and Git on Windows is recalcitrant.

It's probably faster for me to install a Linux VM (another area of darkness for me) and fix it there. But if it's a Git bug, then it should be fixed.

I will try to create a repo here on github that you can clone. Of course you'll have to do the preparations - create a case sensitive parent folder where you can clone the repo into. As I said, my git expertise on a scale of 0 to 100 is around 0.5, so I might need some time ;).

**Comment by Rolf-B** [comment]
Okay, I've done it. Use this repo: <https://github.com/Rolf-B/caseProblem>. It also contains a description of all necessary steps.

**Comment by tboegi** [comment]
Nicely done.
There are some questions:

`git clone "https://github.com/Rolf-B/caseProblem.git" "D:/Projekte/caseProblem"`
Should this read ... D:/Git_CS/caseProblem" ?

Another question, to nail things down a little bit:
What does the following give you under Git bash ?
cd /d/Git_CS
mkdir test_repo
cd test_repo
git init
git config -l core.ignorecase

And what does this give, in the same directory ?
echo FOO >FOO.TXT
echo foo >foo.txt
cat FOO.TXT
cat foo.txt

**Comment by Rolf-B** [comment]
> Should this read ... D:/Git_CS/caseProblem" ?

Ugh, of course. More precisely: D:\Git_CS\caseProblem

> git config -l core.ignorecase

Doesn't work - error: wrong number of arguments, should be 0

Should this read `git config --get core.ignorecase`? 😛 
This says false and `git config --local --get core.ignorecase`  says nothing. Inspecting the contents of .git/config shows no entry for [core] ignorecase.

> And what does this give, in the same directory ?

It gives that FOO.TXT contains FOO and foo.txt contains foo. 

After some tinkering with my caseProblem repo, I seem to have found something. Files that I add on my local machine work fine. I added them from Windows commandline and from git bash, I added them, committed them, pushed them, re-cloned the repo - no problems. Then I added files in GitHub, cloned, and lo and behold - Serverfile.txt (and Foo.txt) is reported as changed after cloning. Bashfile/bashfile and Cmdfile/cmdfile work just fine. I will ask a friend to push two more files from a linux machine.
